java/ql/lib/semmle/code/java/security/Files.qll

```diff
@@ -0,0 +1,72 @@
+      [
+        "java.io;FileOutputStream;false;FileOutputStream;;;Argument[0];create-file",
+        "java.io;RandomAccessFile;false;RandomAccessFile;;;Argument[0];create-file",
+        "java.io;FileWriter;false;FileWriter;;;Argument[0];create-file",
+        "java.io;PrintStream;false;PrintStream;(File);;Argument[0];create-file",
+        "java.io;PrintStream;false;PrintStream;(File,String);;Argument[0];create-file",
+        "java.io;PrintStream;false;PrintStream;(File,Charset);;Argument[0];create-file",
+        "java.io;PrintStream;false;PrintStream;(String);;Argument[0];create-file",
+        "java.io;PrintStream;false;PrintStream;(String,String);;Argument[0];create-file",
+        "java.io;PrintStream;false;PrintStream;(String,Charset);;Argument[0];create-file",
+        "java.io;PrintWriter;false;PrintWriter;(File);;Argument[0];create-file",
+        "java.io;PrintWriter;false;PrintWriter;(File,String);;Argument[0];create-file",
+        "java.io;PrintWriter;false;PrintWriter;(File,Charset);;Argument[0];create-file",
+        "java.io;PrintWriter;false;PrintWriter;(String);;Argument[0];create-file",
+        "java.io;PrintWriter;false;PrintWriter;(String,String);;Argument[0];create-file",
+        "java.io;PrintWriter;false;PrintWriter;(String,Charset);;Argument[0];create-file",
+        "java.nio.file;Files;false;copy;;;Argument[1];create-file",
+        "java.nio.file;Files;false;createDirectories;;;Argument[0];create-file",
+        "java.nio.file;Files;false;createDirectory;;;Argument[0];create-file",
+        "java.nio.file;Files;false;createFile;;;Argument[0];create-file",
+        "java.nio.file;Files;false;createLink;;;Argument[0];create-file",
+        "java.nio.file;Files;false;createSymbolicLink;;;Argument[0];create-file",
+        "java.nio.file;Files;false;createTempDirectory;;;Argument[0];create-file",
+        "java.nio.file;Files;false;createTempFile;(Path,String,String,FileAttribute[]);;Argument[0];create-file",
+        "java.nio.file;Files;false;move;;;Argument[1];create-file",
+        "java.nio.file;Files;false;newBufferedWriter;;;Argument[0];create-file",
+        "java.nio.file;Files;false;newOutputStream;;;Argument[0];create-file",
+        "java.nio.file;Files;false;write;;;Argument[0];create-file",
+        "java.nio.file;Files;false;writeString;;;Argument[0];create-file"
+      ]
+  }
+}
+
+private class WriteFileSinkModels extends SinkModelCsv {
+  override predicate row(string row) {
+    row =
+      [
+        "java.io;FileOutputStream;false;write;;;Argument[0];write-file",
+        "java.io;RandomAccessFile;false;write;;;Argument[0];write-file",
+        "java.io;RandomAccessFile;false;writeBytes;;;Argument[0];write-file",
+        "java.io;RandomAccessFile;false;writeChars;;;Argument[0];write-file",
+        "java.io;RandomAccessFile;false;writeUTF;;;Argument[0];write-file",
+        "java.io;Writer;true;append;;;Argument[0];write-file",
+        "java.io;Writer;true;write;;;Argument[0];write-file",
```

Comment:
Seems inconsistent that Writer.write is modelled as a file-write but FileOutputStream is special-cased rather than inheriting from OutputStream.write?

Reply:
My reasoning for this was that, if constructed properly, several kinds of `Writer` can be used to write to a file, e.g.

```java
FileOutputStream os = new FileOutputStream("some_file.txt");
Writer writer = new BufferedWriter(new OutputStreamWriter(os, "utf-8"));
Writer writer2 = new OutputStreamWriter(os, "utf-8");
```

Whereas, AFAIK, the same cannot be said about `OutputStream`, i.e. it needs to be a `FileOutputStream` if you want to write to a file.

Reply:
Ah sure, I thought you were handling cases where you're not sure where the `OutputStream` came from. Worth checking if we track taint well enough that the typical `BufferedOuputStream(FileOutputStream(...))` works?

Reply:
Hmm you're right, the case for `OutputStream`s is actually very similar to `Writer` since you can always wrap it in another stream.

I don't think it currently works for that case, because `BufferedOutputStream.write` isn't defined as a sink, and AFAIK we don't support this kind of wrapping when calculating virtual dispatch.

So, for consistency, if you agree I could change the `FileOutputStream` sink to `OutputStream`, and see if it generates too many FPs. Because I'm afraid that the opposite (modeling only `FileWriter`) would cause too many FNs.